controller/konnect/constraints.go

```diff
@@ -0,0 +1,42 @@
+package konnect
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configurationv1 "github.com/kong/kubernetes-configuration/api/configuration/v1"
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+)
+
+// SupportedKonnectEntityType is an interface that all Konnect entity types
+// must implement.
+type SupportedKonnectEntityType interface {
+	configurationv1alpha1.KongService |
+		configurationv1alpha1.KongRoute |
+		configurationv1.KongConsumer
+	// TODO: add other types
+
+	GetTypeName() string
+}
+
+// EntityType is an interface that all Konnect entity types must implement.
+// Separating this from SupportedKonnectEntityType allows us to use EntityType
+// where client.Object is required, since it embeds client.Object and uses pointer
+// to refer to the SupportedKonnectEntityType.
+type EntityType[
+	T SupportedKonnectEntityType,
+] interface {
+	*T
+
+	// Kubernetes Object methods
+
+	GetObjectMeta() metav1.Object
+	client.Object
+
+	// Added methods
```

Comment:
Can we set a more meaningful comment here?

Reply:
Done. PTAL.